wsimod/extensions.py

```diff
@@ -0,0 +1,116 @@
+    >>> def patch_t(node):
+    >>>     return 10
+
+Example of patching an attribute item:
+
+`patch_default_pull_set_handler` will be assigned to
+`my_node.pull_set_handler["default"]`:
+
+    >>> @register_node_patch("my_node.pull_set_handler", item="default")
+    >>> def patch_default_pull_set_handler(self, vqip):
+    >>>     return {}
+
+If patching a method of an attribute, the `is_attr` argument should be set to `True` and
+the target should include the node name and the attribute name and the method name, all
+separated by periods, eg. `node_name.attribute_name.method_name`.
+
+It should be noted that the patched function should have the same signature as the
+original method or attribute, and the return type should be the same as well, otherwise
+there will be a runtime error.
+
+Finally, the `apply_patches` is called within the `Model.load` method and will apply all
+patches in the order they were registered. This means that users need to be careful with
+the order of the patches in their extensions files, as they may have interdependencies.
+
+TODO: Update documentation on extensions files.
+"""
+
+from typing import Callable, Hashable
+
+from .orchestration.model import Model
+
+extensions_registry: dict[tuple[str, Hashable, bool], Callable] = {}
+
+
+def register_node_patch(
+    target: str, item: Hashable = None, is_attr: bool = False
+) -> Callable:
+    """Register a function to patch a node method or any of its attributes.
+
+    Args:
+        target (str): The target of the object to patch as a string with the node name
+            attribute, sub-attribute, etc. and finally method (or attribue) to replace,
+            sepparated with period, eg. `node_name.make_discharge` or
+            `node_name.sewer_tank.pull_storage_exact`.
+        item (Hashable): Typically a string or an integer indicating the item to replace
+            in the selected attribue, which should be a list or a dictionary.
+        is_attr (bool): If True, the decorated function will be called when applying
+            the patch and the result assigned to the target, instead of assigning the
+            function itself. In this case, the only argument passed to the function is
+            the node object.
+    """
+    target_id = (target, item, is_attr)
+    if target_id in extensions_registry:
+        raise ValueError(f"Patch for {target} already registered.")
+
+    def decorator(func):
+        extensions_registry[(target, item, is_attr)] = func
+        return func
+
+    return decorator
+
+
+def apply_patches(model: Model) -> None:
+    """Apply all registered patches to the model.
+
+    TODO: Validate signature of the patched methods and type of patched attributes.
+
+    Args:
+        model (Model): The model to apply the patches to.
+    """
+    for (target, item, is_attr), func in extensions_registry.items():
+        # Process the target string
+        starget = target.split(".")
```

Comment:
there isn't anything to stop nodes have a `.` in the node name.. It's not in my default model setup - but I wouldn't be surprised if others have introduced this unthinkingly. 

Perhaps could we have `target_node` as a separate argument - since the `name` could be anything, and then any sub-attributes can by `.` delimited since they will follow python syntax? 

Not sure - what do you think? If it's too awful then at least we validate to ensure no `.` in `name` during `model.load`

Reply:
Not a problem. That's an easy fix. I just put it all together in a single line because I felt it was easier to understand and to cover more cases - in particular the sub-attributes one - in one, consistent approach.

About users using `.` for the node names... it might be a bit opinionated, but you are the developer, so you tell the users how they should use the tool. If you tell them not to use `.` but `_` or something else, they won't use `.`. It is not the other way around.

Reply:
Actually, you're right. There's no point on artificially restricting what a node name can be. What about changing the decorator signature to?

```python
def register_node_patch(
    node_name: str, target: str, item: Hashable = None, is_attr: bool = False
) -> Callable:
```

So `node_name` is provided independently and can therefore be anything?

Reply:
works for me!